Add FITS normalization with astropy.simple_norm and user-specified file list

gbrammer opened the pull request:
This PR provides the option for more custom normalization of FITS images using [astropy.visualization.simple_norm](https://docs.astropy.org/en/stable/visualization/normalization.html).  A normalization object is generated from the image data and keywords specified in `tile_img.norm_kwargs`, which is then passed as `norm` to `convert.make_tile_mpl`.

The second minor change allows the user to specify an explicit list of filenames in `convert.dir_to_map(..., filelist=['file1.fits','file2.png', ...])`, rather than generating the list automatically with `os.listdir`, which provides files in a random order.  The function checks that the requested files exist in the indicated `directory`.

Thanks for providing this great tool!

ryanhausen commented:
Hi @gbrammer!

Thanks for trying out FitsMap and an even bigger thanks for contributing!

re: [astropy.visualization.simple_norm integration](https://github.com/ryanhausen/fitsmap/pull/28/commits/e24f0eaf621844e52450e376ea1cdc9b82aed2eb)
I really like the idea of using astropy.visualization.simple_norm, I have really been wanting a plug-and-play solution for scaling raw fits images and I think this is a natural fit. 

I have a couple suggestions wrt the implementation. If I understand the documentation for [simple_norm](https://docs.astropy.org/en/stable/api/astropy.visualization.mpl_normalize.simple_norm.html) correctly, the default setting is a linear scaling from image_min to image_max (which is the current fitsmap implementation). If that's the case, we can drop the `vmin`/`vmax` settings from [`make_tile_mpl`](https://github.com/gbrammer/fitsmap/blob/5cafaad22b6cdefb332ae4bb1734bd6291745a60/fitsmap/convert.py#L307) and always use a `simple_norm` object. 

These changes will require some tests to be updated because of the change in arguments, some are failing right now. 
<hr/>

re: [using a file_list](https://github.com/ryanhausen/fitsmap/pull/28/commits/d74e3d81840735d4e0299edfed519cd7a810bfc6)
First I need to apologize for poor documentation on my part (especially in the README). I think this functionality is already implemented in the [`convert.files_to_map`](https://github.com/gbrammer/fitsmap/blob/5cafaad22b6cdefb332ae4bb1734bd6291745a60/fitsmap/convert.py#L924), which takes a list of strings which are paths to files. `convert.dir_to_map` is just a convenience function that calls that function. So, in the case where you want to create a map using specific files not a directory, it makes sense to just call `convert.files_to_map`. 

I will update the README to reflect this use case, sorry!
<hr/>

re:[add wcs information to panner](https://github.com/gbrammer/fitsmap/blob/5cafaad22b6cdefb332ae4bb1734bd6291745a60/fitsmap/convert.py#L924)
This seems really neat! 

Do I understand it correctly the ra/dec will be stored in the url and then make particular views reference-able via url?
<hr/>

I am almost done with a pretty big update on the [async-layers-markers](https://github.com/ryanhausen/fitsmap/tree/async-layers-markers) branch. Which has minor python changes, but is almost a complete javascript rewrite. The changes allow the map to load extra tileLayers and catalogs asynchronously. I can help update/merge your changes after that branch is done/merged. Luckily the only major conflict I can see is the cartographer/javascript stuff not the simple_norm. Merging both should be doable with a little hacking.

I put some other comments in the commits themselves.

Thanks again for contributing and I am glad that you are finding FitsMap useful!

ryanhausen commented on fitsmap/cartographer.py:
```diff
@@ -269,9 +275,9 @@ def markers_to_js(marker_file_names: List[str]) -> str:
         "            src = collection[j];",
         "",
-        "            var width = (((src.widest_col * 10) * src.n_cols) + 10).toString() + 'em';",
+        "            var width = '900px';",
```
Did you find that the width and height that were set before were too/small or not the right ratio? I would like to avoid hard-coding if possible.

gbrammer replied:
I didn't mean to commit this with the other changes.  I'm working on a change to revert it to your default and add a global variable to allow it to be modified.

gbrammer commented:
1. I just pushed a few changes to implement your suggestion to always use simple_norm, but with linear min/max scaling by default.  I also updated the tests accordingly, and almost all now pass.  The only thing still failing in the travis build is `test_files_to_map`, which fails for the "td width" thing I removed and tiny floating point differences on the "x" and "y" values in `test_catalog_radec.cat.js`.

2. I didn't actually notice how `dir_to_map` is just a simple wrapper around `files_to_map`.   I just pushed a minor change to sort the `os.listdir` result, since I think that's preferable over the random sorting it gives you otherwise.  

3. The WCS stuff adds a few variables to the javascript pulled from the WCS header, along with functions for doing the (assumed simple) coordinate conversions.  The "window.location" stuff puts the celestial coordinates in the address bar and updates them when the map view changes.  The result is a "permalink" that can then parsed to recover the pan/zoom position.  [Here](https://s3.amazonaws.com/grizli-v1/Pipeline/j142632p3508/Map/index.html?ra=216.6342823&dec=35.1376096&zoom=6) is an example I just compiled.  Getting that to work is why I pulled the `center` value from the `map` initialization.  Even without the WCS, I suppose the same technique could get the x/y pan and zoom.

4. I don't think it should be too much trouble to stick in the WCS javascript if a bunch of other things change with that branch you mentioned.  The only thing I changed in the main javascript is initializing `map` with a single image layer and then adding the others with `layers.control` as suggested [here](https://leafletjs.com/examples/layers-control/).

ryanhausen commented:
@gbrammer I have merged the `async-layers-markers` branch into `main` and am now tweaking the PR to merge successfully. If you haven't already, can you check the box to ["allow edits from maintainers" ](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork). 

Thanks!

ryanhausen commented:
@gbrammer  I have made the edits to your fork so that I think it will merge.  Before I do, would you mind trying it out to see if it still works as expected? I can push the changes to your fork, if you give me permission as outlined [here](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork). Otherwise, I'll merge them into main and we can fix any bugs that come up.

Thanks again for the great contributions!

gbrammer commented:
Thanks, @ryanhausen.  The "allow edits from maintainers" box is already checked.  It may take me a few days to check the new branch.

ryanhausen commented:
Thanks @gbrammer, I pushed some changes to a new [branch](https://github.com/gbrammer/fitsmap/tree/gbrammer/master). Let me know what you think. I haven't edited someone else's PR before, let me know if I edited it in a way that made it hard to use. It may still take a few edits to merge, but functionally, it should contain both your new features and the features I was working on.

ryanhausen commented:
@gbrammer Thanks for your contribution, your merged branch was uploaded to pypi as v0.7.0.

Let me know if you run into any issues!

Thanks again for contributing!
